Mathlib/RingTheory/MvPowerSeries/Substitution.lean

```diff
@@ -0,0 +1,1096 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir, María Inés de Frutos Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos Fernández
+-/
+
+import Mathlib.Algebra.MvPolynomial.CommRing
+import Mathlib.LinearAlgebra.Finsupp.Pi
+import Mathlib.RingTheory.MvPowerSeries.Evaluation
+import Mathlib.RingTheory.MvPowerSeries.LinearTopology
+import Mathlib.RingTheory.MvPowerSeries.Trunc
+import Mathlib.RingTheory.PowerSeries.Basic
+import Mathlib.RingTheory.Nilpotent.Basic
+import Mathlib.Topology.Algebra.Algebra
+
+/-! # Substitutions in power series
+
+Here we define the substitution of power series into other power series.
+We follow Bourbaki, Algèbre, chap. 4, §4, n° 3.
+
+Evaluation of a power series at adequate elements has been defined
+in `DividedPowers.ForMathlib.MvPowerSeries.Evaluation.lean`.
+The goal here is to check the relevant hypotheses:
+* The ring of coefficients is endowed the discrete topology.
+* The main condition rewrites as having vanishing constant coefficient
+* Power series have a linear topology
+-/
+
+theorem Set.Finite.support_of_summable
+    {α : Type*} [AddCommGroup α] [TopologicalSpace α] [DiscreteTopology α]
+    {β : Type*} (f : β → α) (h : Summable f) : Set.Finite f.support := by
+  obtain ⟨a, ha⟩ := h
+  simp only [HasSum] at ha
+  classical
+  simp_rw [tendsto_atTop_nhds] at ha
+  obtain ⟨s, hs⟩ := ha {a} rfl (isOpen_discrete _)
+  apply Set.Finite.subset s.finite_toSet
+  intro b
+  rw [Function.mem_support, not_imp_comm]
+  intro hb
+  let hs' := hs (insert b s) (s.subset_insert b)
+  simp only [Set.mem_singleton_iff] at hs hs'
+  simpa [Finset.sum_insert hb, hs, add_left_eq_self] using hs'
+
+theorem IsNilpotent.finsum {α : Type*} [CommSemiring α] {β : Type*} {f : β → α}
+    (hf : ∀ b, IsNilpotent (f b)) :
+    IsNilpotent (finsum f) := by
+  classical
+  by_cases h : Set.Finite f.support
+  · rw [finsum_def, dif_pos h]
+    exact Commute.isNilpotent_sum (fun b _ ↦ hf b) (fun i j hi hj ↦ Commute.all _ _)
+  · simp only [finsum_def, dif_neg h, IsNilpotent.zero]
```

Comment:
```suggestion
  rw [finsum_def]
  split_ifs with h
  · exact Commute.isNilpotent_sum (fun b _ ↦ hf b) (fun i j hi hj ↦ Commute.all _ _)
  · exact IsNilpotent.zero
```

Reply:
This should go right next to [isNilpotent_sum](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Nilpotent/Basic.html#isNilpotent_sum), and have a `Commute` version like [Commute.isNilpotent_sum](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Nilpotent/Basic.html#Commute.isNilpotent_sum).